File: diffguard/src/main/java/com/diffguard/agent/strategy/StrategyPlanner.java
package com.diffguard.agent.strategy;

import com.diffguard.agent.strategy.ReviewStrategy.AgentType;
import com.diffguard.model.DiffFileEntry;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

/**
 * 审查策略规划器。
 * <p>
 * 分析 diff 内容，自动选择最优审查策略：
 * - 确定启用哪些专用 Agent
 * - 设置各 Agent 的权重
 * - 添加针对性的审查规则
 * <p>
 * 策略选择逻辑：
 * - Controller 变更 → 重点 API 兼容性 + 安全
 * - DAO 变更 → 重点 SQL 风险 + 性能
 * - Service 变更 → 重点业务逻辑 + 性能
 * - Config 变更 → 重点安全配置
 * - 安全敏感代码 → 安全 Agent 权重最高
 * - 数据库操作 → 安全 + 性能 Agent
 * - 并发代码 → 性能 Agent 权重最高
 */
public class StrategyPlanner {

    private static final Logger log = LoggerFactory.getLogger(StrategyPlanner.class);

    /**
     * 分析 diff 并生成审查策略。
     */
    public static ReviewStrategy plan(List<DiffFileEntry> diffEntries) {
        DiffProfile profile = DiffProfiler.profile(diffEntries);
        ReviewStrategy strategy = selectStrategy(profile);

        log.info("策略规划完成: {} (风险={}, 主要类别={}, 启用Agent={})",
                strategy.getName(),
                profile.getOverallRisk(),
                profile.getPrimaryCategory(),
                strategy.getEnabledAgents());

        return strategy;
    }

    /**
     * 获取 diff 画像（用于调试或日志）。
     */
    public static DiffProfile getProfile(List<DiffFileEntry> diffEntries) {
        return DiffProfiler.profile(diffEntries);
    }

    private static ReviewStrategy selectStrategy(DiffProfile profile) {
        ReviewStrategy.Builder builder = ReviewStrategy.builder()
                .name("auto-detected")
                .priority(profile.getOverallRisk().ordinal());

        // 默认全部启用
        builder.agentWeight(AgentType.SECURITY, 1.0);
        builder.agentWeight(AgentType.PERFORMANCE, 1.0);
        builder.agentWeight(AgentType.ARCHITECTURE, 1.0);

        DiffProfile.FileCategory primary = profile.getPrimaryCategory();

        switch (primary) {
            case CONTROLLER -> applyControllerStrategy(builder, profile);
            case DAO -> applyDaoStrategy(builder, profile);
            case SERVICE -> applyServiceStrategy(builder, profile);
            case CONFIG -> applyConfigStrategy(builder, profile);
            case MODEL -> applyModelStrategy(builder, profile);
            case TEST -> applyTestStrategy(builder, profile);
            default -> applyDefaultStrategy(builder, profile);
        }

        // 全局风险调整
        adjustForRisk(builder, profile);

        return builder.build();
    }

    private static void applyControllerStrategy(ReviewStrategy.Builder builder, DiffProfile profile) {
        builder.name("API 兼容性审查")
                .description("Controller 变更：重点检查 API 兼容性、参数验证、权限控制");

        builder.agentWeight(AgentType.SECURITY, 1.5);
        builder.agentWeight(AgentType.ARCHITECTURE, 1.3);

        builder.focusArea("API 接口兼容性：URL 路径、HTTP 方法、参数是否向后兼容");
        builder.focusArea("参数验证：@RequestBody/@PathVariable 是否有校验");
        builder.focusArea("权限控制：接口是否有正确的权限注解");
        builder.focusArea("异常处理：全局异常处理是否覆盖新增异常类型");

        if (profile.hasExternalApiCalls()) {
            builder.additionalRule("重点检查 API 契约变更，确认不影响现有客户端");
        }
    }

    private static void applyDaoStrategy(ReviewStrategy.Builder builder, DiffProfile profile) {
        builder.name("数据库安全审查")
                .description("DAO 变更：重点检查 SQL 安全、查询性能、事务管理");

        builder.agentWeight(AgentType.SECURITY, 2.0);
        builder.agentWeight(AgentType.PERFORMANCE, 1.5);
        builder.agentWeight(AgentType.ARCHITECTURE, 0.5);

        builder.focusArea("SQL 注入风险：是否使用参数化查询");
        builder.focusArea("查询性能：是否有全表扫描、缺少索引的查询");
        builder.focusArea("事务管理：@Transactional 配置是否正确");
        builder.focusArea("数据完整性：外键约束、唯一约束是否满足");

        if (profile.hasDatabaseOperations()) {
            builder.additionalRule("所有 SQL 语句必须使用参数化查询，禁止字符串拼接");
            builder.additionalRule("检查是否有批量操作可能导致锁表");
        }
    }

    private static void applyServiceStrategy(ReviewStrategy.Builder builder, DiffProfile profile) {
        builder.name("业务逻辑审查")
                .description("Service 变更：重点检查业务逻辑正确性、并发安全、异常处理");

        builder.agentWeight(AgentType.SECURITY, 1.0);
        builder.agentWeight(AgentType.PERFORMANCE, 1.3);
        builder.agentWeight(AgentType.ARCHITECTURE, 1.0);

        builder.focusArea("业务逻辑正确性：边界条件、空值处理、异常分支");
        builder.focusArea("事务一致性：跨表操作是否在事务中");

        if (profile.hasConcurrencyCode()) {
            builder.agentWeight(AgentType.PERFORMANCE, 2.0);
            builder.focusArea("并发安全：共享状态是否有正确的同步机制");
            builder.additionalRule("检查是否有竞态条件和死锁风险");
        }

        if (profile.hasDatabaseOperations()) {
            builder.focusArea("N+1 查询：循环中是否有数据库调用");
        }
    }

    private static void applyConfigStrategy(ReviewStrategy.Builder builder, DiffProfile profile) {
        builder.name("安全配置审查")
                .description("配置变更：重点检查安全配置、密钥管理、环境差异");

        builder.agentWeight(AgentType.SECURITY, 2.5);
        builder.agentWeight(AgentType.PERFORMANCE, 0.3);
        builder.agentWeight(AgentType.ARCHITECTURE, 0.5);

        builder.focusArea("硬编码密钥：配置文件中是否有明文密码或 API Key");
        builder.focusArea("环境差异：开发/测试/生产配置是否一致");
        builder.focusArea("权限配置：是否有过于宽松的权限设置");
        builder.additionalRule("禁止在配置文件中硬编码密码、密钥等敏感信息");
    }

    private static void applyModelStrategy(ReviewStrategy.Builder builder, DiffProfile profile) {
        builder.name("数据模型审查")
                .description("Model 变更：重点检查序列化、向后兼容、验证注解");

        builder.agentWeight(AgentType.SECURITY, 0.8);
        builder.agentWeight(AgentType.PERFORMANCE, 0.5);
        builder.agentWeight(AgentType.ARCHITECTURE, 1.5);

        builder.focusArea("序列化兼容：新增/删除字段是否影响 JSON 序列化");
        builder.focusArea("验证注解：字段是否有正确的 @NotNull/@Size 等注解");
        builder.focusArea("向后兼容：字段变更是否影响现有 API 契约");
    }

    private static void applyTestStrategy(ReviewStrategy.Builder builder, DiffProfile profile) {
        builder.name("测试质量审查")
                .description("测试变更：降低审查强度，重点检查测试覆盖度");

        builder.agentWeight(AgentType.SECURITY, 0.3);
        builder.agentWeight(AgentType.PERFORMANCE, 0.3);
        builder.agentWeight(AgentType.ARCHITECTURE, 0.5);

        builder.focusArea("测试覆盖：变更是否被充分测试");
        builder.focusArea("测试质量：断言是否充分，边界条件是否覆盖");
    }

    private static void applyDefaultStrategy(ReviewStrategy.Builder builder, DiffProfile profile) {
        builder.name("通用代码审查")
                .description("通用审查：全面检查安全、性能和架构问题");

        builder.focusArea("代码质量：命名规范、复杂度、重复代码");
        builder.focusArea("错误处理：异常是否被正确处理");

        if (profile.hasDatabaseOperations()) {
            builder.agentWeight(AgentType.SECURITY, 1.5);
        }
        if (profile.hasConcurrencyCode()) {
            builder.agentWeight(AgentType.PERFORMANCE, 1.5);
        }
    }

    private static void adjustForRisk(ReviewStrategy.Builder builder, DiffProfile profile) {
        switch (profile.getOverallRisk()) {
            case CRITICAL -> {
                builder.additionalRule("高风险变更：所有 Agent 必须完成完整审查");
                builder.priority(3);
            }
            case HIGH -> {
                builder.additionalRule("中高风险变更：建议启用全部审查");
                builder.priority(2);
            }
            case MEDIUM -> builder.priority(1);
            case LOW -> {
                // 低风险可选择性减少审查
                builder.priority(0);
            }
        }
    }
}

File: services/agent/app/agent/strategy_planner.py
"""Diff profiler and strategy planner - ported from Java."""

from __future__ import annotations

import enum
import re
from dataclasses import dataclass, field

from app.models.schemas import DiffEntry


class FileCategory(enum.Enum):
    CONTROLLER = "controller"
    SERVICE = "service"
    DAO = "dao"
    MODEL = "model"
    CONFIG = "config"
    TEST = "test"
    UTILITY = "utility"
    OTHER = "other"


class RiskLevel(enum.Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class AgentType(enum.Enum):
    SECURITY = "security"
    PERFORMANCE = "performance"
    ARCHITECTURE = "architecture"


@dataclass
class DiffProfile:
    total_files: int = 0
    total_lines: int = 0
    category_distribution: dict[FileCategory, int] = field(default_factory=dict)
    has_database_operations: bool = False
    has_security_sensitive_code: bool = False
    has_concurrency_code: bool = False
    has_external_api_calls: bool = False
    overall_risk: RiskLevel = RiskLevel.LOW


@dataclass
class ReviewStrategy:
    name: str = "default"
    description: str = ""
    agent_weights: dict[AgentType, float] = field(default_factory=dict)
    focus_areas: list[str] = field(default_factory=list)
    additional_rules: list[str] = field(default_factory=list)

    def get_enabled_agents(self) -> list[AgentType]:
        return [at for at, w in self.agent_weights.items() if w > 0]

    def get_enabled_agent_names(self) -> list[str]:
        """Return agent names as strings (for AgentRegistry lookup)."""
        return [at.value for at, w in self.agent_weights.items() if w > 0]


# --- Pattern matching ---

_FILE_CATEGORY_PATTERNS: list[tuple[re.Pattern, FileCategory]] = [
    (re.compile(r"(Controller|Resource|Endpoint|Api)\.java$", re.I), FileCategory.CONTROLLER),
    (re.compile(r"(Service|ServiceImpl|Manager|Handler|Processor)\.java$", re.I), FileCategory.SERVICE),
    (re.compile(r"(Repository|Dao|Mapper|Repo)\.java$", re.I), FileCategory.DAO),
    (re.compile(r"(Entity|Model|Dto|Vo|Request|Response|Bean)\.java$", re.I), FileCategory.MODEL),
    (re.compile(r"(Config|Configuration|Properties|Settings)\.java$", re.I), FileCategory.CONFIG),
    (re.compile(r"(Test|Spec|IT)\.java$", re.I), FileCategory.TEST),
    (re.compile(r"(Util|Utils|Helper|Constants|Converter)\.java$", re.I), FileCategory.UTILITY),
]

_DB_PATTERNS = re.compile(
    r"(JdbcTemplate|PreparedStatement|ResultSet|@Query|@Entity|@Table|"
    r"INSERT|UPDATE|DELETE|SELECT\s|executeQuery|executeUpdate|createQuery|"
    r"save\(|findById|findAll|findBy)",
    re.I,
)

_SECURITY_PATTERNS = re.compile(
    r"(@PreAuthorize|@Secured|@RolesAllowed|Authentication|Principal|"
    r"Password|Secret|Token|JWT|OAuth|Cookie|Session|"
    r"encrypt|decrypt|hash|cipher|XSS|CSRF|injection|sanitize)",
    re.I,
)

_CONCURRENCY_PATTERNS = re.compile(
    r"(synchronized|volatile|AtomicInteger|AtomicLong|ConcurrentHashMap|"
    r"@Lock|ReentrantLock|Semaphore|CountDownLatch|CompletableFuture|"
    r"ThreadPool|ExecutorService|@Async|Thread\.|runnable)",
    re.I,
)

_API_PATTERNS = re.compile(
    r"(@RequestMapping|@GetMapping|@PostMapping|@PutMapping|@DeleteMapping|"
    r"RestTemplate|WebClient|HttpClient|OkHttp|Feign|fetch\(|axios|"
    r"ResponseBody|RequestBody|@PathVariable|@RequestParam)",
    re.I,
)


def profile(entries: list[DiffEntry]) -> DiffProfile:
    """Analyze diff entries to produce a DiffProfile."""
    p = DiffProfile(total_files=len(entries))
    categories: dict[FileCategory, int] = {}

    for entry in entries:
        content = entry.content
        p.total_lines += content.count("\n")

        # Classify file
        cat = FileCategory.OTHER
        for pattern, category in _FILE_CATEGORY_PATTERNS:
            if pattern.search(entry.file_path):
                cat = category
                break
        categories[cat] = categories.get(cat, 0) + 1

        # Detect keywords
        if _DB_PATTERNS.search(content):
            p.has_database_operations = True
        if _SECURITY_PATTERNS.search(content):
            p.has_security_sensitive_code = True
        if _CONCURRENCY_PATTERNS.search(content):
            p.has_concurrency_code = True
        if _API_PATTERNS.search(content):
            p.has_external_api_calls = True

    p.category_distribution = categories

    # Risk assessment
    risk_score = 0
    if p.has_database_operations:
        risk_score += 2
    if p.has_security_sensitive_code:
        risk_score += 2
    if p.has_concurrency_code:
        risk_score += 1
    if p.has_external_api_calls:
        risk_score += 1
    if p.total_files > 10:
        risk_score += 1

    if risk_score >= 5:
        p.overall_risk = RiskLevel.CRITICAL
    elif risk_score >= 3:
        p.overall_risk = RiskLevel.HIGH
    elif risk_score >= 1:
        p.overall_risk = RiskLevel.MEDIUM
    else:
        p.overall_risk = RiskLevel.LOW

    return p


def _primary_category(profile: DiffProfile) -> FileCategory:
    if not profile.category_distribution:
        return FileCategory.OTHER
    return max(profile.category_distribution, key=profile.category_distribution.get)


class StrategyPlanner:
    """Selects a review strategy based on the diff profile.

    Strategy weights and focus areas are loaded from ``strategy/config.yaml``
    when available, with fallback to built-in defaults.
    """

    def __init__(self) -> None:
        self._config = _load_strategy_config()

    def plan(self, entries: list[DiffEntry]) -> ReviewStrategy:
        p = profile(entries)
        primary = _primary_category(p)

        # Start from defaults
        weights: dict[AgentType, float] = {
            AgentType(name): w
            for name, w in self._config.defaults.items()
        }

        # Ensure all agent types present
        for at in AgentType:
            if at not in weights:
                weights[at] = 1.0

        focus: list[str] = []
        rules: list[str] = []

        # Apply category-specific overrides from config
        cat_key = primary.value
        if cat_key in self._config.categories:
            cat_cfg = self._config.categories[cat_key]
            for agent_name, weight in cat_cfg.weights.items():
                at = AgentType(agent_name)
                weights[at] = weight
            focus.extend(cat_cfg.focus)
        else:
            # Fallback: hardcoded category rules
            focus.extend(_builtin_category_focus(primary))
            _apply_builtin_category_weights(primary, weights)

        # Adjust for risk level
        if p.overall_risk in (RiskLevel.HIGH, RiskLevel.CRITICAL):
            risk_key = "high_and_critical"
            risk_adjs = self._config.risk_adjustments.get(risk_key, {})
            if p.has_database_operations and "has_database_operations" in risk_adjs:
                adj = risk_adjs["has_database_operations"]
                weights[AgentType.SECURITY] = weights.get(AgentType.SECURITY, 1.0) + adj.security_delta
                focus.extend(adj.focus)
            elif p.has_database_operations:
                weights[AgentType.SECURITY] = weights.get(AgentType.SECURITY, 1.0) + 0.5
                focus.append("Database operation safety")

            if p.has_concurrency_code and "has_concurrency_code" in risk_adjs:
                adj = risk_adjs["has_concurrency_code"]
                weights[AgentType.PERFORMANCE] = weights.get(AgentType.PERFORMANCE, 1.0) + adj.performance_delta
                focus.extend(adj.focus)
            elif p.has_concurrency_code:
                weights[AgentType.PERFORMANCE] = weights.get(AgentType.PERFORMANCE, 1.0) + 0.5
                focus.append("Thread safety and concurrency")

        return ReviewStrategy(
            name=f"{primary.value}_review",
            description=f"Primary category: {primary.value}, Risk: {p.overall_risk.value}",
            agent_weights=weights,
            focus_areas=focus,
            additional_rules=rules,
        )


# --- Config loader ---


def _load_strategy_config():
    """Load strategy config from YAML. Returns a StrategyConfig even on failure."""
    try:
        from app.agent.strategy.config_loader import StrategyConfig
        return StrategyConfig.load()
    except Exception:
        return _DefaultConfig()


class _DefaultConfig:
    """Fallback config when YAML loading fails."""
    defaults = {"security": 1.0, "performance": 1.0, "architecture": 1.0}
    categories = {}
    risk_adjustments = {}


# --- Built-in fallback rules (used when YAML config is absent) ---


def _builtin_category_focus(category: FileCategory) -> list[str]:
    return {
        FileCategory.CONTROLLER: ["Input validation", "Authentication/Authorization", "API design"],
        FileCategory.DAO: ["SQL injection", "Query performance", "Transaction handling"],
        FileCategory.SERVICE: ["Business logic correctness", "Error handling"],
        FileCategory.CONFIG: ["Configuration security", "Secret exposure"],
        FileCategory.MODEL: ["Data integrity", "Serialization safety"],
        FileCategory.TEST: ["Test coverage", "Test quality"],
    }.get(category, [])


def _apply_builtin_category_weights(category: FileCategory, weights: dict[AgentType, float]) -> None:
    overrides = {
        FileCategory.CONTROLLER: {AgentType.SECURITY: 1.5, AgentType.ARCHITECTURE: 1.3},
        FileCategory.DAO: {AgentType.SECURITY: 2.0, AgentType.PERFORMANCE: 1.5, AgentType.ARCHITECTURE: 0.5},
        FileCategory.SERVICE: {AgentType.PERFORMANCE: 1.3},
        FileCategory.CONFIG: {AgentType.SECURITY: 2.5, AgentType.PERFORMANCE: 0.3},
        FileCategory.MODEL: {AgentType.ARCHITECTURE: 1.5, AgentType.SECURITY: 0.8},
        FileCategory.TEST: {AgentType.SECURITY: 0.3, AgentType.PERFORMANCE: 0.3, AgentType.ARCHITECTURE: 0.5},
    }.get(category, {})
    weights.update(overrides)
